### DISCUSSION TITLE
next/image: Provide a more flexible layout option

### DISCUSSION THREAD

**Discussion by amuttsch** [vercel/next.js#21588] [Ideas]
**Describe the feature you'd like to request**

Currently, the available options for image sizes are fixed, intrinsic, responsive and fill. In our case, none of those options fit our needs as we display book covers which vary in aspect ratio ever so slightly.

We display the images in a CSS grid which uses auto columns for sizing and all content inside should fill to take its available space. Because of this we also need an image that fills to its available space. The provided "fill" layout should handle this use case according to the docs. But because of our component elements, which are mostly width and height `100%`, the image collapses because of the `position: absolute` and there is no defined size to fill to.

The solution that would work for us is to use a native browser `img` tag with `width: 100%` and `height: auto`.

Here is an example of how next/image handles our images in the given context vs. how we want the images to behave (below the horizontal rule): https://jsfiddle.net/wLfyc7xu/1/

Notice that the nextjs images using layout fill on top are completely hidden and how the bottom images scale when you shrink the width of your browser.

**Describe the solution you'd like**

Previously there was an `unsized` attribute that would give us the flexibility to style and size the images to our requirements: https://github.com/vercel/next.js/commit/9569f5a31ac627f271260ab2f30572036ce3fa5b

It would be great if you could reintroduce this attribute and maybe just return the `img` without the surrounding `div`.

Another viable option for us would be if the `generateImgAttrs` function would be exported, that way we could build our own `img` tag from scratch. Or a function that returns just the `srcset` for a given image url.

**Describe alternatives you've considered**

The other image layouts do not work because we do not know the image sizes and we loop over images which have slightly different sizes. We also tried to work around the given styling with no success, it is too restrictive for our use case.

**Comment by ReangeloJ** [comment]
I would like to see this too!
Ive resorted to juat using the native img for time being. Hopefully this will be solved later on!

**Comment by samparmenter** [comment]
Yeah, this is a bit of a glaring omission from the image component. We have a lot of images on our site that need to conform to a given space which is almost never full width. The image component seems like its designed around a set of specific use cases and doesn't provide access to the internals that would allow us to extend or make use of the great tech under the hood.

**Comment by akshaysmurthy** [comment]
This is the most frustrating bit about `next/image`. For us too, in many cases we wouldn't know what the aspect ratio of the images would be. 

@amuttsch Did you happen to find out anything more re this?

**Reply by amuttsch** [reply]
In the end we implemented our own image component that uses the next image optimization backend. So unfortunately no, we weren't able to do this with the provided `next/image`.

**Reply by akshaysmurthy** [reply]
Makes sense, but that's also an interesting point you brought up. I hope you don't mind me going slightly off topic here - I understand that the backend follows a certain URL structure that makes it possible to use the image optimisation even without using the component. But this is not documented anywhere, right? And would love your thoughts on https://github.com/vercel/next.js/discussions/23298 :)

**Reply by amuttsch** [reply]
It is documented in the code here https://github.com/vercel/next.js/blob/canary/packages/next/client/image.tsx#L719 ;) I guess this url won't change in the near future, but we also implemented our own image optimization backend because of https://github.com/vercel/next.js/issues/23637 :-)

**Reply by adambraimbridge** [reply]
> In the end we implemented our own image component that uses the next image optimization backend. So unfortunately no, we weren't able to do this with the provided `next/image`.

@amuttsch thanks for this well-written issue. Any chance we can see the image component you wrote? Cheers :+1:

**Comment by stubar** [comment]
For me I'm finding the current layout options aren't sufficient for a fluid grid layout regardless if I know the aspect ratio or not.  The problem stems from all decisions about which image size to download being based in the width of the entire window.  We have images that maybe fullwidth, 3 abreast, 4 abreast, all based on a grid.  It seems the recommended solution is to use `sizes` but I can't see how this can account for grid gutters & an overall site max width, it would be much easier to say base your image choice on the clientWidth of the image.  https://stackoverflow.com/questions/61296655/how-to-serve-image-size-based-on-container-not-screen-size This details the problem you can have with fluid images.  An answer suggests using calc but it's a pain to convert grids/margins/max layout widths into calcs for every new image layout you build (a limitation of html5 spec not next btw).  I'm planning to build a wrapper around the image that senses the client width and sets the width/height props of next/image accordingly, it will also override certain Next styles with !important.

**Reply by samparmenter** [reply]
I've had to write my own wrapper around the url based image optimisations nextjs offer for exactly this reason. On mobile most of our images are 100% width or close to. On desktop very few of them are so we are requesting very large images for the sake of a 200px wide image. I can't imagine that most websites have only full width images regardless of viewport so its strange this isn't taken into consideration.

**Reply by stubar** [reply]
I am currently weighing up the pros and cons of a similar approach.  At my last company before Next/Image was around we used a Sanity back end for images which has a great image processing back end API.   I was using window.ResizeObserver to sense the width of the image and set the src accordingly.  It also had base64 processing like Next/Image (albeit switching the src out with an onLoad event).  The only downside I am starting to realise with this, is the server can never know what the clientWidths are of the image in a fluid layout because it is unaware of the browser width.  So if I were to view the server rendered page with JS turned off, I would only see the base64 representations.

My current line of thinking is to go in extreme on the sizes parameter, our images are only ever going to occupy x number of columns in our grid layout and what their widths are at anyone one time is a function of the vw, overall max site width, number of columns & grid gutter size.  If I call this function for every image's sizes prop, passing in a breakpoint to columns map e.g. `{base:1, md:2, lg 16}` it should be able to give back a string of media queries and their respective css calc function.  Admittedly writing the function for the site's grid system is a royal PITA (but you should only have to write it once per site) and likewise it requires far more thought when creating new image templates.

**Reply by stubar** [reply]
Just to follow up on this I have implemented my idea of using function to write the sizes parameter in a fluid grid with breakpoints and have had success.  I'm convinced this is the best approach now.
